src/mca/ptl/base/ptl_base_fns.c

```diff
@@ -1458,33 +1469,47 @@ char **pmix_ptl_base_split_and_resolve(char **orig_str, char *name)
 
         /* Go through all interfaces and see if we can find a match */
         for (if_index = pmix_ifbegin(); if_index >= 0; if_index = pmix_ifnext(if_index)) {
-            pmix_ifindextoaddr(if_index, (struct sockaddr *) &if_inaddr, sizeof(if_inaddr));
-            if (pmix_net_samenetwork(&argv_inaddr, &if_inaddr, argv_prefix)) {
-                break;
+            pmix_ifindextoaddr(if_index,
+                               (struct sockaddr*) &if_inaddr,
+                               sizeof(if_inaddr));
+            if (pmix_net_samenetwork((struct sockaddr*) &argv_inaddr,
+                                     (struct sockaddr*) &if_inaddr,
+                                     argv_prefix)) {
+                /* We found a match. If it's not already in the interfaces array,
+                   add it. If it's already in the array, treat it as a match */
+                match_count = match_count + 1;
+                pmix_ifindextoname(if_index, if_name, sizeof(if_name));
+                for (n = 0; n < interface_count; n++) {
+                    if (0 == strcmp(if_name, interfaces[n])) {
+                        break;
+                    }
+                }
+                if (n == interface_count) {
+                    pmix_output_verbose(20,
+                                        pmix_ptl_base_framework.framework_output,
+                                        "ptl:tool: Found match: %s (%s)",
+                                        pmix_net_get_hostname((struct sockaddr*) &if_inaddr),
+                                        if_name);
+                    pmix_argv_append(&interface_count, &interfaces, if_name);
+                }
             }
+
         }
         /* If we didn't find a match, keep trying */
-        if (if_index < 0) {
+        if (0 == match_count) {
             pmix_show_help("help-ptl-base.txt", "invalid if_inexclude", true, name, tmp,
                            "Did not find interface matching this subnet");
             free(tmp);
             continue;
         }
-
-        /* We found a match; get the name and replace it in the
-           argv */
-        pmix_ifindextoname(if_index, if_name, sizeof(if_name));
-        pmix_output_verbose(20, pmix_ptl_base_framework.framework_output,
-                            "ptl:tool: Found match: %s (%s)",
-                            pmix_net_get_hostname((struct sockaddr *) &if_inaddr), if_name);
-        argv[save++] = strdup(if_name);
-        free(tmp);
     }
 
-    /* The list may have been compressed if there were invalid
-       entries, so ensure we end it with a NULL entry */
-    argv[save] = NULL;
+    /* Mark the end of the interface name array with NULL */
+    if (NULL != interfaces) {
+        interfaces[interface_count] = NULL;
+    }
+    pmix_argv_free(argv);
     free(*orig_str);
-    *orig_str = pmix_argv_join(argv, ',');
-    return argv;
+    *orig_str = pmix_argv_join(interfaces, ',');
```

Comment:
I don't understand this statement - isn't the `orig_str` the list of interfaces to include/exclude? You definitely do _not_ want to free and overwrite that info as those are MCA params - this is likely to cause the param system to segfault upon clean termination.

Reply:
My understanding is that in the original code, argv had the mca parameter list specifying the interfaces to be included/excluded. 

The original code performed an in-place update of the passed argv array, either keeping the original specification if it was an interface name, and replacing it with the first matching interface if it was a subnet specification. There's some validation of the parameters, so in the original code, invalid parameters are skipped, meaning the resulting argv array may be shortened, so a null is appended following the last replacement argv to terminate the array at that point.

The original code then freed whatever we in orig_str and replacing it by building a comma delimited string of whatever is in the new, updated argv array.

My changes were intended to replicate this functionality, but now since a subnet specification can expand to one or more interface names, a new array containing the final list of interfaces to be returned is needed.

It seemed I was making this work as it did originally, but now returning multiple interface names for a subnet spec. 

I could be missing something, but my tests have been successful with this change.

Reply:
To clarify, orig_str was what had the original -mca parameter list amd the argv array was created from that by splitting at commas. But still, in the original code, orig_str was freed and replaced at the end of the function.